packages/workspaces-extension/src/top_indicator.tsx

```diff
@@ -0,0 +1,139 @@
+    const handleClickOutside = (event: MouseEvent) => {
+      if (
+        dropdownRef.current &&
+        !dropdownRef.current.contains(event.target as Node)
+      ) {
+        setIsExpanded(false);
+      }
+    };
+
+    document.addEventListener('mousedown', handleClickOutside);
+    return () => document.removeEventListener('mousedown', handleClickOutside);
+  }, []);
+
+  return (
+    <div className="jp-WorkspaceSelector" ref={dropdownRef}>
+      <button
+        className="jp-WorkspaceSelectorHeader"
+        onClick={() => setIsExpanded(!isExpanded)}
+        aria-expanded={isExpanded}
+      >
+        <span className="jp-WorkspaceSelectorCurrent">
+          {currentWorkspace.length > 12
+            ? `${currentWorkspace.slice(0, 12)}...`
+            : currentWorkspace}
+        </span>
+        <span className="jp-WorkspaceSelectorCaret">
+          {isExpanded ? '▲' : '▼'}
+        </span>
+      </button>
+
+      {isExpanded && (
+        <div className="jp-WorkspaceSelectorDropdown">
+          <div className="jp-WorkspaceSelectorSearch">
+            <div className="jp-WorkspaceSelectorSearchIcon">
+              <LabIcon.resolveReact icon={searchIcon} />
+            </div>
+            <input
+              type="text"
+              className="jp-WorkspaceSelectorInput"
+              placeholder="Search"
+              value={searchQuery}
+              onChange={e => setSearchQuery(e.target.value)}
+              autoFocus
+            />
+          </div>
+
+          <ul className="jp-WorkspaceSelectorList">
+            {filteredIdentifiers.map(identifier => (
+              <li
+                key={identifier}
+                className="jp-WorkspaceSelectorItem"
+                onClick={() => {
+                  if (identifier === currentWorkspace) return;
+                  openWorkspace(identifier);
+                  setIsExpanded(false);
+                }}
+              >
+                <LabIcon.resolveReact
+                  icon={identifier === currentWorkspace ? checkIcon : blankIcon}
+                />
+                {identifier.length > 12
+                  ? `${identifier.slice(0, 12)}...`
+                  : identifier}
+              </li>
+            ))}
+          </ul>
+        </div>
+      )}
+    </div>
+  );
+};
+
+/**
+ * A Widget for Workspace Selector at topbar
+ */
+export class WorkspaceSelectorWidget extends ReactWidget {
+  private currentWorkspace: string;
+  private identifiers: string[];
+  private openWorkspace: (workspace: string) => void;
+
+  constructor(props: IWorkspaceSelectorProps) {
+    super();
+    this.id = 'jp-workspace-top-indicator';
+    this.addClass('jp-react-widget');
+    this.identifiers = props.identifiers;
+    this.openWorkspace = props.openWorkspace;
+    this.currentWorkspace = props.currentWorkspace;
+    props.model?.refreshed.connect(() => {
+      this.identifiers = props.model?.identifiers!;
+      this.update();
+    });
```

Comment:
Let's not introduced any non-null assertions unless in test. Here is appears unnecessary at a first glance.

```suggestion
    props.model.refreshed.connect(() => {
      this.identifiers = props.model.identifiers;
      this.update();
    });
```